### ISSUE TITLE
Memory allocation patterns

### ISSUE THREAD

**Issue by mikepurvis** [cparse/cparse#64]
My application is pre-compiling symbolic expressions in `calculator` instances and then repeatedly eval'ing them at runtime. Obviously, the compilation stage allocates lots of TokenXX objects as it builds up the RPN list, but it looks like it then goes ahead and allocates essentially a parallel set on each evaluation:

https://github.com/cparse/cparse/blob/a9f26860283c2653f8a0fcb003cbb07f21b690af/shunting-yard.cpp#L494-L502

For an application where memory allocation (or at least syscalls) need to be avoided at runtime, it'd be great to have the expression evaluator able to work without making new allocations (or perhaps with the option of a pre-allocated value stack for the RPN engine).

**Comment by VinGarcia** [comment]
Hello @mikepurvis, I think allocations would be needed to store intermediary values when processing the RPN, but your suggestion about a preallocated value stack seems a nice way to go about it.

However, being honest, this might be a little trick to solve in the short term.

The easiest solution would be to preallocate an array of bytes and work on them for all new allocations.
However, I don't know yet how to control the allocation of some of the types;
for example, some of the types are implemented internally using the C++ containers `std::string` and `std::map` and although I know I can use allocators to control how the memory management of these vectors work I have never played with this concept.

You might know more about these details than I do and I will welcome any advice.

I can also share with you an idea I have (for some moment in the future) to implement these Tokens using a union of one of two words of 64 bits.
I know JavaScript works with a single word of 64bits for each variable, but it also has (or had) some problems for allocating doubles, so it might be worth considering using more space but having a cleaner implementation.
It would still require some allocation mechanism for lists, maps and big strings but it would allow us more control over how the memory is used.

So I will think about what I can do, if you have any suggestions please say so.

**Comment by VinGarcia** [comment]
Also, are you using lists and maps or only basic types such as numbers and booleans? We might also consider thinking a solution that solves your use case first and waiting for making a bigger change with more time.

**Comment by mikepurvis** [comment]
I'm not using any assignment operators at all— it's purely just crunching through pre-compiled expressions like `x + 1.23 * y + 123.0`, where the TokenMap supplying those vars has been cooked in advance, and is just having its valued changed. So I'm not needing to expand any vectors or maps on the fly.

I suspect the biggest issue here is that the the RPN stack is not uniform in type, so it really does depend on having abstract TokenBase pointers to heap-allocated objects. It's possible the only way to make this work would be to have pools of each different kind of object, which can be called upon during evaluation.

Alternatively, it's possible my use case is niche enough that I should just write my own evaluator which bakes in the assumptions I can make— critically that all symbols, expressions, and functions resolve down to a double. Then I can use a fixed-size stack of type double and just perform the evaluations against that. I'd still get the enormous benefit of cparse's ability to parse to RPN, but I'd have more control over the evaluation side.

**Comment by VinGarcia** [comment]
Yes this is an option that might work and should'nt be too complicated once you already have the RPN ready :thinking:.

I have started working on that plan I suggested of imitating the JavaScript Token structure, if I have a working version I will share it with you, it would allow us to have tokens with the same size, i.e.:

```c++
union tokenData_t {
  uint64_t integer;
  double real;

  // For more complex types:
  void* pointer;
};
```

If I have any good results until this weekend you might be able to use it.

**Comment by VinGarcia** [comment]
@mikepurvis I did not have enough time to implement this feature, you probably should try that solution using you suggested building an evaluator that receives the RPN as input and assumes everything evaluates to float.

Also if you have time and implement something you think would fit CParse you are welcome to submit a PR.

**Comment by jpfeger** [comment]
I have run the Visual Studio performance profiler on the library for the same medium sized formula calculated 1000 times. The results show that
1 - 'delete' operator takes up to 25.5 % of the inclusive time while 'new' accounts for only 1.3 %. Assuming that most of the objects built are Tokens, improvement could be then gained from using a pool of these instances
2 - there is a significant amount of object instances built for each calculation (150+). This comes as a suprise. Not sure what conclusion to draw from that neither which fix to apply if any
3 - regular STL operations on Tree, deque are not very time consuming (fewer than 10% of total time). Sounds normal to me.

 I was able to lower the a calculation time for a specific formula from 44 micro-seconds down to 32 micro by using references whenever possible. Strangely enough passing Config_t per value was the main killer. I could push my improvements and profiler results if interest is shown.

**Comment by VinGarcia** [comment]
Closing this issue, for now, it's possible to optimize this by implementing type-specific arrays (like an int array) so that instead of having several tokens per list we'll keep a single token containing an `std::vector<int>` or something like that. But this would need to be done for all types, or at least all basic types. If anyone wants to try this it's okay and we can reopen this ticket, but I imagine this won't be easy.
